### PR TITLE
Rewrite flush logic

### PR DESCRIPTION
**Pull request by ralphtheninja**
Closes https://github.com/Level/level-ws/issues/10, closes https://github.com/Level/level-ws/issues/59

### CODE REVIEW

File: level-ws.js [hunk 1]
```diff
@@ -69,62 +73,58 @@ inherits(WriteStream, Writable)
...
 
+  // TODO remove, no longer needed
   if (!self._db.isOpen()) {
     return self._db.once('ready', function () {
       write.call(self, d, enc, next)
     })
   }
 
-  if (self._options.maxBufferLength &&
-      self._buffer.length > self._options.maxBufferLength) {
```
Comment (ralphtheninja) [inline level-ws.js]: This commit temporarily removes support for `maxBufferLength`. Focus was to remove race conditions first.

File: level-ws.js [hunk 2]
```diff
@@ -69,62 +73,58 @@ inherits(WriteStream, Writable)
...
-    self._buffer.push(d)
-    next()
-  }
+  // TODO re-implement buffering logic using this._options.maxBufferLength
+  self._buffer.push(d)
+  next()
 }
 
-WriteStream.prototype._flush = function (f) {
+WriteStream.prototype._final = function (cb) {
```
Comment (ralphtheninja) [inline level-ws.js]: According to documentation, implementations of `WriteStream` should implement `_write()` and optionally `_final()`.

File: test.js [hunk 3]
```diff
@@ -394,3 +394,24 @@ test('test that missing type errors', function (t, ctx, done) {
   ws.write(data)
   ws.end()
 })
+
+test('batches according to maxBufferLength', function (t, ctx, done) {
```
Comment (ralphtheninja) [inline test.js]: Currently we have `sourceData.length === 10`, which means we will batch `5` times with each batch having `2` items.

File: level-ws.js [hunk 4]
```diff
@@ -83,12 +83,27 @@ WriteStream.prototype._write = function write (d, enc, next) {
...
+  } else {
+    push(d)
+  }
 }
 
 WriteStream.prototype._final = function (cb) {
+  this._batch(cb)
+}
+
+WriteStream.prototype._batch = function (cb) {
```
Comment (ralphtheninja) [inline level-ws.js]: Basically renamed `_final()` to `_batch()` and let `_final()` call `_batch()`, `_write()` then handles buffering and `maxBufferLength`.

File: level-ws.js [hunk 5]
```diff
@@ -84,7 +84,15 @@ WriteStream.prototype._write = function write (d, enc, next) {
   }
 
   function push (d) {
-    self._buffer.push(d)
+    self._buffer.push({
```
Comment (ralphtheninja) [inline level-ws.js]: If we push transformed objects directly to `_buffer` there's no need to `.map()` to create a new array.

**Comment by ralphtheninja** [comment]
`$ time run-for -n 10000`

With `standard` disabled.

screenshot from 2018-06-26 01-58-32

**Comment by vweevers** [comment]
So this:

- No longer batches everything written in a tick
- Changes the meaning of `maxBufferLength` (before, anything over this limit would be ignored) (now, it controls when the buffer is written to the db)
- If `options.maxBufferLength` is not defined, no `db.batch()` will happen until the end of the stream.

@ralphtheninja is that right? If so, I think this changes too much.

**Comment by ralphtheninja** [comment]
I misunderstood maxBufferLength. Will come up with something else.

**Comment by vweevers** [comment]
@ralphtheninja `maxBufferLength` is kind of a dangerous option though. Maybe we should remove it? I don't see why one would ever want that behavior. I prefer having backpressure - i.e. having `write` return false if the buffer is too big.

**Comment by vweevers** [comment]
Sorry, I mixed up readable and writable streams there (with returning false), but I hope you get my point.

**Comment by ralphtheninja** [comment]
@vweevers I'm at a point where I'm about to give up _again_ :smile: I'm fairly certain that it's the `process.nextTick()` in `_write` that messes this up.

**Comment by vweevers** [comment]
I can help out fulltime on friday ❤️ 

Until then, an idea: write a unit test that monkey-patches `process.nextTick` (not globally, only in `write`) to do `setTimeout(fn, 1000)`. That might trigger the race condition?

**Comment by ralphtheninja** [comment]
I'll do some simple cleanup stuff before so the code is easier to work with.

**Pull request closed without merging**